### DISCUSSION TITLE
Collaboration between Vike and Lit

### DISCUSSION THREAD

**Discussion by brillout** [lit/lit#4076] [General]
Author here of [vite-plugin-ssr](https://vite-plugin-ssr.com/) and Vike (the upcoming rebrand of vite-plugin-ssr).

Some updates on Vike and why I believe it would make sense for us to collaborate in a deep fashion.

## Long-term vision

Beyond technical considerations, there is one aspect I'd like to emphasize: the long-term vision of Vike which is to become an open foundation to build upon.

Everything about Vike has been designed around extensibility and lower-level primitives. For example, Vike is the only framework that can deeply integrate with Relay. The long-term goal being to be able to build all kinds of frameworks such as Gatsby's data layer or Astro's island architecture.

## No vendor lock in

We believe the fact that Vercel, Next.js, and React are getting increasingly tightly coupled with each other is a terrible outcome for users.

Users want freedom, and rightfully so: being vender-locked is a major threat for companies. Most companies are aware of that.

Vike unlocks a new kind of freedom: users can choose React, Lit, Vue, or whatever UI framework they want.

Telling users "pick Lit because of its merits" is an appealing story.

## Being able to use Lit while tapping into React's ecosystem

A major USP is that users can use both Lit and React for the same app.

Lit can be used for most pages, while React can be used only for a couple of pages that need libraries that only exist for React.

This also means that users can more seamlessly migrate between React and Lit.

The story being: "Try Lit and if you don't like it you can fallback to React or Vue. Although we believe you'll end up liking Lit a lot!"

## Built-in Router

Thanks to Vike's built-in rotuer, Vike brings a *lot* more features to the table than alternatives that are router agnostic.

The goal of the [V1 release](https://github.com/brillout/vite-plugin-ssr/issues?q=is%3Aissue+is%3Aopen+label%3A%22v1+release+%3Astar2%3A%22) is to be feature complete. The router will include all features users may expect: type safe routes, route guards, URL rewriting, being able to define all routes in a Single Route File, ...

All the Lit team has to do is to implement the Lit integration. Enabling Lit's team to focus on what makes Lit special.

## Vike roadmap

 - [V1 release](https://github.com/brillout/vite-plugin-ssr/issues?q=is%3Aissue+is%3Aopen+label%3A%22v1+release+%3Astar2%3A%22)
 - Vike rebrand. With a marketing that targets a wider audience with a more professional/dry tone. (The vite-plugin-ssr copy has more of a personal/hobbyist/passionate tone.)

Once the V1 is released, Vike can be considered feature complete, highly polished, and robust. AFAICT the V1 alone is enough to grow to 100k, but we'll also invest in marketing to grow beyond 100k.

From a user yesterday:

 > I appreciate the descriptive and helpful error messages of vite-plugin-ssr more than once.

That's just the tip of the iceberg of the amount of meticulous care put into details. The V1 release will be the pinnacle of that.

## Conclusion

What are your toughts? Looking forward to discuss this.

**Comment by e111077** [comment]
Just wondering, are there any roadblocks for integrating Lit? From what I've seen, like with the e.g. [the Astro integration](https://github.com/withastro/astro/blob/main/packages/integrations/lit/server.js) it's pretty straightforward and only about 150 lines.

We unfortunately won't be able to dedicate resources to this project, but we can help ensure that there is nothing preventing a Vike integration from our side. I very much like the ideals of framework / platform agnosticism as it lines up much with the project's goals and capabilities already and I think that you might be able to recruit some like-minded individuals to help you out with a Lit integration on [our discord](https://lit.dev/discord).

**Reply by justinfagnani** [reply]
I think pitching this and recruiting on our Discord sounds like a great idea. Hopefully our SSR APIs are straightforward enough that most people could handle it, and I think we could help guide and review.

**Reply by brillout** [reply]
I was told Lit is considering building a meta framework (but I don't really know as I'm not familiar with Lit).

I wanted to reach out as Vike's goal very much is to enable folks like yourself to build meta frameworks.

> it's pretty straightforward

It should also be straightforward with Vike then. (In general, if a tool can be integrated with Astro, then it most likely can with Vike as well since Vike is more flexible.) 

Also, Vike is very much open to create new low-level primitives to further enable deep integration, if that's something you'd end up needing.

Thanks for your work on Lit, it looks like a very refreshing and innovative alternative.
